### DISCUSSION TITLE
If I compile the firmware myself, machine.Timer crashes the board.

### DISCUSSION THREAD

**Discussion by allstar87** [micropython/micropython#11948] [ESP32]
First, I would like to express a general frustration with the state of documentation surrounding MicroPython in general and it's ESP32 port in specific.  Everything from the built-in module APIs to the instructions for compiling the firmware are terse to the point of incompleteness.  I get the feeling I'm supposed to already know everything I'm trying to learn from the official docs, and I almost always have to google for old forum threads or random tutorials for the necessary detail and context.

I'm only interested in compiling the firmware myself because the dev boards I prefer working with--the Sparkfun Thing Plus--comes with 16MB of Flash instead of 4, and no ready-compiled version of the firmware is compiled for 16MB of flash.

I managed to compile a version of the firmware, essentially using the included partitions_16MiB.csv file to make my own SF_THING_PLUS board profile, otherwise copy-pasted from GENERIC.  It has been working for me for days, I've been messing with Wifi and MQTT and such, decided to try out the machine.Timer functions and any attempt at setting a timer interrupt crashes the board with a Guru Meditation Error.  I built the GENERIC firmware, none of my changes.  Erased the flash and flashed that to the board.  Same problem.  I download the pre-compiled version from micropython.org, install that with esptool.py...it works fine.  

Troubleshooting this one is beyond my level of skill and understanding; why would compiling the firmware myself be a problem?

**Comment by shariltumin** [comment]
I do not think the problem was due to compiling your own firmware. The machine.timer module seems to be broken with firmware compiled with ESP-IDF-5.0.2, previous firmware compiled with ESP-IDF-4.x.x was fine.

I am sure that the MicroPython core developers are working hard to fix some issues caused by the migration from ESP-IDF-4.x.x to ESP-IDF-5.x.x since the ESP-IDF-4 is no longer supported.


```
$ mv ~/Downloads/esp32-20230627-unstable-v1.20.0-261-g813d559bc.bin firmware.bin

$ strings firmware.bin | grep IDF
MicroPython-1.20.0-xtensa-IDFv5.0.2-with-newlib4.1.0

>>> 
MPY: soft reboot
MicroPython v1.20.0-261-g813d559bc on 2023-06-27; ESP32 module with ESP32
Type "help()" for more information.
>>> from machine import Timer
>>> tim0 = Timer(0)
>>> tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))
>>> Guru Meditation Error: Core  1 panic'ed (Interrupt wdt timeout on CPU1). 

```

----------------------------------------------------------------------------

```
$ strings firmware.bin-502 | grep IDF
MicroPython-1.20.0-xtensa-IDFv5.0.2-with-newlib4.1.0

MPY: soft reboot
MicroPython v1.20.0-242-g6a9db521e on 2023-07-06; ESP32 module with ESP32
Type "help()" for more information.
>>> from machine import Timer
>>> tim0 = Timer(0)
>>> tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))
D (78872) intr_alloc: Connected src 14 to int 9 (cpu 1)
>>> Guru Meditation Error: Core  1 panic'ed (Interrupt wdt timeout on CPU1). 

```

---------------------------------------------------------------------------

```
$ strings firmware.bin | grep IDF
MicroPython-1.20.0-xtensa-IDFv4.4.5-with-newlib3.3.0


MPY: soft reboot
MicroPython v1.20.0-206-g33b403dfb-kaki5 on 2023-07-06; ESP32 module with ESP32
Type "help()" for more information.

>>> 
>>> from machine import Timer
>>> tim0 = Timer(0)
>>> tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))
>>> 0

>>> tim1 = Timer(1)
>>> tim1.init(period=2000, mode=Timer.PERIODIC, callback=lambda t:print(1))
>>> 1
1
1
1
...

```

**Reply by allstar87** [reply]
Okay yeah that's consistent with what I'm seeing.  So my compiled firmware is busted but it's nothing I did "wrong."

I don't really need to use that library; I don't have a project at the moment, I'm just playing and learning and seeing what the board--and MicroPython--can do.

**Comment by scruss** [comment]
>... I download the pre-compiled version from micropython.org, install that with esptool.py...it works fine.

For anyone inclined to try, esptool's `--dont-append-digest`  and `--flash_size 16MB` options for the *elf2image* and *write_flash* commands don't seem to allow an elf image to be reformatted to a bin file that understands the Thing+ ESP32's 16 MB chip. Using `-e` with *write_flash* does detect and erase all 16 MB (slowly), but a modified bin image gets stuck in a boot loop.

Trying to force a 4 MB image to use all available flash does write a working MicroPython image, but all the extra storage is ignored:
```
esptool.py --chip esp32 --port /dev/tty.usbserial-14130  --baud 460800 write_flash -e  --flash_size detect -z 0x1000 esp32-20230426-v1.20.0.bin
esptool.py v4.6.2
Serial port /dev/tty.usbserial-14130
Connecting....
Chip is ESP32-D0WD-V3 (revision v3.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 94:b9:7e:79:c5:78
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Auto-detected Flash size: 16MB
Erasing flash (this may take a while)...
Chip erase completed successfully in 46.4s
Warning: Image file at 0x1000 is protected with a hash checksum, so not changing the flash size setting. Use the --flash_size=keep option instead of --flash_size=16MB in order to remove this warning, or use the --dont-append-digest option for the elf2image command in order to generate an image file without a hash checksum
Compressed 1566528 bytes to 1034676...
Wrote 1566528 bytes (1034676 compressed) at 0x00001000 in 24.9 seconds (effective 503.7 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```
There's a very strong chance I'm doing something wrong, though

**Comment by adrianlzt** [comment]
@shariltumin I have also compiled esp32 with idf v5.0.2 ([instructions](https://github.com/adrianlzt/micropython-esp32-cam-docker) and algo get an error with the timer (using Lilygo camera v1.6.2):
```
MicroPython v1.20.0-261-g813d559bc on 2023-07-07; ESP32-cam module (i2s) with ESP32
Type "help()" for more information.
>>> from machine import Timer
>>> tim0 = Timer(0)
>>> tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))
>>> :0x3ffbc860x40097e6f:0x3ffbc820 0x40149086:0x3ffbc840 0x400d397f0 CP0f91fc  PS      : 0x00050035  A0      : 0x40082ec8  A1      : 0x3ffc04a0
A2      : 0x3f80d530  A3      : 0x00000000  A4      : 0x00050021  A5      : 0x40099eda
A6      : 0x3ffc40a8  A7      : 0x3ffc5210  A8      : 0x3ff5f000  A9      : 0x00000000:0x3ffbc860 0x401e4687:0x3ffbc880


Core  0 register dump:
PC      : 0x40097e72  PS      : 0x00060c35  A0      : 0x80149089  A1      : 0x3ffbc220
A2      : 0x00000003  A3      : 0x00060c23  A4      : 0x00060c20  A5      : 0x3ffb6ed0
A6      : 0x00000003  A7      : 0x00060023  A8      : 0x80090e52  A9      : 0x3ffbc200
A10     : 0x3ffc40a4  A11     : 0x00000000  A12     : 0x3ffbe774  A13     : 0x00000008
A14     : 0x000000a0  A15     : 0x00060c23  SAR     : 0x0000001d  EXCCAUSE: 0x00000006
EXCVADDR: 0x00000000  LBEG    : 0x00000000  LEND    : 0x00000000  LCOUNT  : 0x00000000


Backtrace: 0x40097e6f:0x3ffbc220 0x40149086:0x3ffbc240 0x400d397f:0x3ffbc260 0x401e4687:0x3ffbc280




ELF file SHA256: 22ad83ad789ffc7e

Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4672
load:0x40078000,len:14536
load:0x40080400,len:3364
entry 0x400805cc
MicroPython v1.20.0-261-g813d559bc on 2023-07-07; ESP32-cam module (i2s) with ESP32
Type "help()" for more information.
>>>
```

**Comment by robert-hh** [comment]
One observation: I added a pin toggle to the timer ISR function. When the timer fires, this pin toggle fires over and over again for about 300ms, >86000 events in total,  until the crash happens.

**Reply by robert-hh** [reply]
it seems like the interface to the ESP32 lib has to be rewritten. The API calls used in machine_timer.c are not listed any more in the v5.x documentation. The code still builds, but obviously does not work any more.

**Reply by dpgeorge** [reply]
`machine.Timer` was fixed in 7d66ae603d91157069d7cdee065919a61f678fb9

**Reply by robhamerling** [reply]
The quick reference for the ESP32 shows an example with Timer(1) and PERIODIC, but there is no mention of and exception for the ESP32-C3. With my ESP32-C3 (a XIAO and a MINI, firmware
_Micropython v1.22.0-preview.147.ga00c9d56d_) the example code works only partially.  With Timer(0) and Timer(2) it's OK, but with Timer(1) and Timer(3) nothing is ever printed.

**Comment by shariltumin** [comment]
I can confirm that the machine.Timer is now working. Thanks @dpgeorge.

```
MicroPython v1.20.0-283-g7d66ae603 on 2023-07-14; ESP32 DEBUG with ESP32
Type "help()" for more information.
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== from machine import Timer
=== 
=== tim0 = Timer(0)
=== tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))
=== 
D (237691) intr_alloc: Connected src 14 to int 9 (cpu 1)
>>> 0

>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== tim1 = Timer(1)
=== tim1.init(period=2000, mode=Timer.PERIODIC, callback=lambda t:print(1))
=== 
D (280191) intr_alloc: Connected src 15 to int 12 (cpu 1)
>>> 1
1
1
<ctrl>+d
```

Nothing related, but something you might want to know. Three warnings under boot.

```
W (1063) i2s(legacy): legacy i2s driver is deprecated, please migrate to use driver/i2s_std.h, driver/i2s_pdm.h or driver/i2s_tdm.h
W (1081) rmt(legacy): legacy driver is deprecated, please migrate to `driver/rmt_tx.h` and/or `driver/rmt_rx.h`
W (1097) ADC: legacy driver is deprecated, please migrate to `esp_adc/adc_oneshot.h
```
